File: sdks/typescript/packages/observability/src/tracer.ts
import type { PaymentEvent, PaymentEventType } from "./types";

/**
 * Represents a tracked payment flow with its events.
 */
export interface PaymentFlow {
  paymentId: string;
  startTime: number;
  endTime?: number;
  success?: boolean;
  events: PaymentEvent[];
  metadata?: Record<string, unknown>;
}

/**
 * Tracks individual payment flows from start to completion.
 *
 * Automatically computes durations between consecutive steps
 * and provides per-flow querying.
 */
export class PaymentTracer {
  private flows: Map<string, PaymentFlow> = new Map();

  /**
   * Start tracking a new payment flow.
   *
   * @param paymentId - Unique identifier for this payment flow
   * @param metadata - Optional metadata to attach to the flow
   * @returns The created PaymentFlow
   * @throws Error if a flow with this ID already exists
   */
  startFlow(paymentId: string, metadata?: Record<string, unknown>): PaymentFlow {
    if (this.flows.has(paymentId)) {
      throw new Error(`Flow already exists: ${paymentId}`);
    }

    const flow: PaymentFlow = {
      paymentId,
      startTime: Date.now(),
      events: [],
      metadata,
    };

    this.flows.set(paymentId, flow);
    return flow;
  }

  /**
   * Record a step in an existing payment flow.
   *
   * Automatically computes the duration from the previous event.
   *
   * @param paymentId - The payment flow ID
   * @param type - The event type for this step
   * @param data - Optional additional event data
   * @returns The recorded PaymentEvent
   * @throws Error if the flow does not exist
   */
  recordStep(
    paymentId: string,
    type: PaymentEventType,
    data?: Partial<Omit<PaymentEvent, "type" | "timestamp" | "paymentId" | "durationMs">>,
  ): PaymentEvent {
    const flow = this.flows.get(paymentId);
    if (!flow) {
      throw new Error(`Flow not found: ${paymentId}`);
    }

    const now = Date.now();
    const lastEvent = flow.events[flow.events.length - 1];
    const durationMs = lastEvent ? now - lastEvent.timestamp : now - flow.startTime;

    const event: PaymentEvent = {
      type,
      timestamp: now,
      paymentId,
      durationMs,
      ...data,
    };

    flow.events.push(event);
    return event;
  }

  /**
   * Mark a payment flow as completed or failed.
   *
   * @param paymentId - The payment flow ID
   * @param success - Whether the flow completed successfully
   * @param error - Error message if the flow failed
   * @returns The final PaymentEvent (completed or failed)
   * @throws Error if the flow does not exist
   */
  endFlow(paymentId: string, success: boolean, error?: string): PaymentEvent {
    const flow = this.flows.get(paymentId);
    if (!flow) {
      throw new Error(`Flow not found: ${paymentId}`);
    }

    flow.endTime = Date.now();
    flow.success = success;

    const type: PaymentEventType = success ? "payment.completed" : "payment.failed";
    const lastEvent = flow.events[flow.events.length - 1];
    const durationMs = lastEvent ? flow.endTime - lastEvent.timestamp : flow.endTime - flow.startTime;

    const event: PaymentEvent = {
      type,
      timestamp: flow.endTime,
      paymentId,
      durationMs,
      error,
    };

    flow.events.push(event);
    return event;
  }

  /**
   * Get a payment flow by ID.
   *
   * @param paymentId - The payment flow ID
   * @returns The PaymentFlow, or undefined if not found
   */
  getFlow(paymentId: string): PaymentFlow | undefined {
    return this.flows.get(paymentId);
  }

  /**
   * Get all tracked payment flows.
   *
   * @returns Array of all PaymentFlows
   */
  getAllFlows(): PaymentFlow[] {
    return Array.from(this.flows.values());
  }

  /**
   * Get all events for a payment flow in chronological order.
   *
   * @param paymentId - The payment flow ID
   * @returns Array of events, or empty array if flow not found
   */
  getFlowEvents(paymentId: string): PaymentEvent[] {
    const flow = this.flows.get(paymentId);
    return flow ? [...flow.events] : [];
  }

  /**
   * Get the total duration of a completed flow in milliseconds.
   *
   * @param paymentId - The payment flow ID
   * @returns Duration in ms, or undefined if flow not found or not ended
   */
  getFlowDuration(paymentId: string): number | undefined {
    const flow = this.flows.get(paymentId);
    if (!flow || flow.endTime === undefined) {
      return undefined;
    }
    return flow.endTime - flow.startTime;
  }

  /**
   * Remove a tracked flow.
   *
   * @param paymentId - The payment flow ID
   * @returns true if the flow was removed
   */
  removeFlow(paymentId: string): boolean {
    return this.flows.delete(paymentId);
  }

  /**
   * Clear all tracked flows.
   */
  clear(): void {
    this.flows.clear();
  }

  /**
   * Get the number of tracked flows.
   */
  get size(): number {
    return this.flows.size;
  }
}

File: sdks/go/observability/tracer.go
package observability

import (
	"fmt"
	"sync"
	"time"
)

// Flow represents an end-to-end payment flow comprising multiple events.
type Flow struct {
	PaymentID string
	StartTime time.Time
	EndTime   time.Time
	Events    []PaymentEvent
	Success   bool
	Error     string
}

// DurationMs returns the total flow duration in milliseconds, or 0 if incomplete.
func (f *Flow) DurationMs() int64 {
	if f.EndTime.IsZero() || f.StartTime.IsZero() {
		return 0
	}
	return f.EndTime.Sub(f.StartTime).Milliseconds()
}

// Tracer tracks payment flows across multiple steps and records events into a Collector.
type Tracer struct {
	mu        sync.RWMutex
	flows     map[string]*Flow
	collector *Collector
}

// NewTracer creates a Tracer backed by the given Collector.
// If collector is nil, events are tracked but not recorded to a Collector.
func NewTracer(collector *Collector) *Tracer {
	return &Tracer{
		flows:     make(map[string]*Flow),
		collector: collector,
	}
}

// StartFlow begins tracking a new payment flow.
// Returns an error if a flow with the same paymentID already exists.
func (t *Tracer) StartFlow(paymentID string) error {
	if paymentID == "" {
		return fmt.Errorf("observability: paymentID must not be empty")
	}

	t.mu.Lock()
	defer t.mu.Unlock()

	if _, exists := t.flows[paymentID]; exists {
		return fmt.Errorf("observability: flow %q already exists", paymentID)
	}

	now := time.Now()
	flow := &Flow{
		PaymentID: paymentID,
		StartTime: now,
	}
	t.flows[paymentID] = flow

	event := PaymentEvent{
		Type:      EventRequested,
		Timestamp: now,
		PaymentID: paymentID,
	}
	flow.Events = append(flow.Events, event)

	if t.collector != nil {
		t.collector.Record(event)
	}

	return nil
}

// RecordStep adds an intermediate event to an existing flow.
// Returns an error if the flow does not exist or has already ended.
func (t *Tracer) RecordStep(paymentID string, eventType PaymentEventType, data PaymentEvent) error {
	t.mu.Lock()
	defer t.mu.Unlock()

	flow, exists := t.flows[paymentID]
	if !exists {
		return fmt.Errorf("observability: flow %q not found", paymentID)
	}
	if !flow.EndTime.IsZero() {
		return fmt.Errorf("observability: flow %q already ended", paymentID)
	}

	data.Type = eventType
	data.PaymentID = paymentID
	if data.Timestamp.IsZero() {
		data.Timestamp = time.Now()
	}

	flow.Events = append(flow.Events, data)

	if t.collector != nil {
		t.collector.Record(data)
	}

	return nil
}

// EndFlow marks a payment flow as completed or failed.
// Returns the final flow state.
func (t *Tracer) EndFlow(paymentID string, success bool, errMsg string) (*Flow, error) {
	t.mu.Lock()
	defer t.mu.Unlock()

	flow, exists := t.flows[paymentID]
	if !exists {
		return nil, fmt.Errorf("observability: flow %q not found", paymentID)
	}
	if !flow.EndTime.IsZero() {
		return nil, fmt.Errorf("observability: flow %q already ended", paymentID)
	}

	now := time.Now()
	flow.EndTime = now
	flow.Success = success
	flow.Error = errMsg

	eventType := EventCompleted
	if !success {
		eventType = EventFailed
	}

	event := PaymentEvent{
		Type:       eventType,
		Timestamp:  now,
		PaymentID:  paymentID,
		DurationMs: flow.DurationMs(),
		Error:      errMsg,
	}
	flow.Events = append(flow.Events, event)

	if t.collector != nil {
		t.collector.Record(event)
	}

	return flow, nil
}

// GetFlow returns a copy of the flow for the given paymentID.
// Returns nil if the flow does not exist.
func (t *Tracer) GetFlow(paymentID string) *Flow {
	t.mu.RLock()
	defer t.mu.RUnlock()

	flow, exists := t.flows[paymentID]
	if !exists {
		return nil
	}

	// Return a copy to avoid races.
	cp := *flow
	cp.Events = make([]PaymentEvent, len(flow.Events))
	copy(cp.Events, flow.Events)
	return &cp
}

// ActiveFlows returns the number of flows that have not yet ended.
func (t *Tracer) ActiveFlows() int {
	t.mu.RLock()
	defer t.mu.RUnlock()

	count := 0
	for _, f := range t.flows {
		if f.EndTime.IsZero() {
			count++
		}
	}
	return count
}
